engines/private/private.cpp

```diff
@@ -0,0 +1,1120 @@
+#include "private/grammar.tab.h"
+#include "private/grammar.h"
+
+
+namespace Private {
+
+PrivateEngine *g_private = NULL;
+
+extern int parse(char*);
+
+PrivateEngine::PrivateEngine(OSystem *syst, const ADGameDescription *gd)
+    : Engine(syst), _gameDescription(gd) {
+    _rnd = new Common::RandomSource("private");
+
+    g_private = this;
+
+    _saveGameMask = NULL;
+    _loadGameMask = NULL;
+
+    _dossierNextSuspectMask = NULL;
+    _dossierPrevSuspectMask = NULL;
+
+    _nextSetting = NULL;
+    _currentSetting = NULL;
+    _nextMovie = NULL;
+    _nextVS = NULL;
+    _modified = false;
+    _mode = -1;
+    _toTake = false;
+    _repeatedMovieExit = new Common::String("");
+    _pausedSetting = NULL;
+
+
+    // Interface
+    _frame = new Common::String("inface/general/inface2.bmp");
+
+    // Policebust
+    _policeBustEnabled = false;
+    _policeBustSetting = NULL;
+    _numberClicks = 0;
+    policeVideoIndex = 0;
+    _sirenSound = new Common::String("po/audio/posfx002.wav");
+
+
+    // General sounds
+    _globalAudioPath = new Common::String("global/audio/");
+    _noStopSounds = false;
+
+    // Radios and phone
+    _policeRadioArea = NULL;
+    _AMRadioArea = NULL;
+    _phoneArea = NULL;
+
+    // TODO: use this as a default sound for radio
+    _infaceRadioPath = new Common::String("inface/radio/");
+    //_radioSound = new Common::String("inface/radio/radio.wav");
+    //_AMRadioPrefix = new Common::String("inface/radio/comm_/");
+    //_policeRadioPrefix = new Common::String("inface/radio/police/");
+    _phonePrefix = new Common::String("inface/telephon/");
+    _phoneCallSound = new Common::String("phone.wav");
+
+    // Dossiers
+    _dossierPage = 0;
+    _dossierSuspect = 0;
+
+    // Diary
+    _diaryLocPrefix = new Common::String("inface/diary/loclist/");
+
+}
+
+PrivateEngine::~PrivateEngine() {
+    debug("PrivateEngine::~PrivateEngine");
+
+    // Dispose your resources here
+    delete _rnd;
+
+    // Remove all of our debug levels here
+    DebugMan.clearAllDebugChannels();
+}
+
+void PrivateEngine::initializePath(const Common::FSNode &gamePath) {
+    SearchMan.addDirectory(gamePath.getPath(), gamePath, 0, 10);
+}
+
+Common::Error PrivateEngine::run() {
+
+    assert(_installerArchive.open("SUPPORT/ASSETS.Z"));
+    Common::SeekableReadStream *file = NULL;
+
+    // if the full game is used
+    if (!isDemo()) {
+        assert(_installerArchive.hasFile("GAME.DAT"));
+        file = _installerArchive.createReadStreamForMember("GAME.DAT");
+    } else {
+        // if the demo from archive.org is used
+        if (_installerArchive.hasFile("GAME.TXT"))
+            file = _installerArchive.createReadStreamForMember("GAME.TXT");
+
+        // if the demo from the full retail CDROM is used
+        else if (_installerArchive.hasFile("DEMOGAME.DAT"))
+            file = _installerArchive.createReadStreamForMember("DEMOGAME.DAT");
+    }
+
+    assert(file != NULL);
+    void *buf = malloc(file->size()+1);
```

Comment:
It is better to write `char *buf = (char *)malloc(file->size() + 1);` instead of doing the type conversion later.

Reply:
Fixed